tarfs.go

```diff
@@ -40,6 +40,15 @@ type FileSystem struct {
 // a non-empty string, the resulting FileSystem will first look for resources in
 // the local file system before returning an embedded resource.
 func New(tarData []byte, local string) (*FileSystem, error) {
+	return NewWithListener(tarData, local, func(name string, file []byte) []byte {
```

Comment:
I think the most important thing to get right here is the API. I asume `name` is the filename, and `file` the contents, right?
Questions:

- would the `[]byte` argument force the memory to be retained or the file read ahead than it's necessary? Could a reader make more sense potentially?
- is there any other data that could be useful, like file flags, type of file?
- the return type is also `[]byte`. Does that imply that what the listener does is filtering the file in some fashion? Could it be interesting to provide some other value, such as error or a generic interface for metadata?

These are just questions, since I'm missing a lot of context here. If this is solving a very specific issue, and this is the simplest interface for it, it looks good to me.

Reply:
Regarding the first point, it seems that we anyway expand the tar in memory, so never mind about the reader bit.

Reply:
> I think the most important thing to get right here is the API. I asume name is the filename, and file the contents, right?

Yes!

> Questions:
> 
> would the []byte argument force the memory to be retained or the file read ahead than it's necessary? Could a reader make more sense potentially?

The `[]byte` is just the content of the file itself. In our case, we're regexing through that replacing certain paths in CSS files. It looks like we could use a `RuneReader` for that, but doesn't seem like much of a win: https://golang.org/pkg/regexp/

> is there any other data that could be useful, like file flags, type of file?

The callback code doesn't know anything more really.

> the return type is also []byte. Does that imply that what the listener does is filtering the file in some fashion? Could it be interesting to provide some other value, such as error or a generic interface for metadata?

Yeah the calling code is basically filtering it. I don't know what metadata it would return really, and at least currently there are no error cases.

> These are just questions, since I'm missing a lot of context here. If this is solving a very specific issue, and this is the simplest interface for it, it looks good to me.

Yeah we're really just using it for the locally-served UI -- filtering CSS and swapping URL paths to fonts and images with  paths including our secure path to local resources.

My tendency with this stuff is to not over-design it off the bat. If more generalization is necessary, it's easy to add, but generalization itself adds a layer of conceptual complexity in terms of users of the API having to look at docs and the code itself. The simpler it is, the more self-documenting it is IMHO.

Reply:
I agree completely. Thanks for the explanation!